File: src/services/TravelService.cs
using StarSmugglerGo.Domain;
using System;
using System.Collections.Generic;
using System.Linq;

namespace StarSmugglerGo.Services;

public sealed class TravelService
{
    public int GetTravelCost(PortDefinition fromPort, PortDefinition toPort)
    {
        if (fromPort.Id == toPort.Id)
        {
            return 0;
        }

        int baseCost = 15;
        int zoneDifference = Math.Abs((int)fromPort.Zone - (int)toPort.Zone);
        int cost = baseCost + (zoneDifference * 2);

        if (zoneDifference >= 2)
        {
            cost *= 2;
        }

        return cost;
    }

    public int GetCheapestTravelCostFromPort(PortDefinition origin, IEnumerable<PortDefinition> ports)
    {
        int cheapest = int.MaxValue;

        foreach (PortDefinition destination in ports)
        {
            if (destination.Id == origin.Id)
            {
                continue;
            }

            cheapest = Math.Min(cheapest, GetTravelCost(origin, destination));
        }

        return cheapest == int.MaxValue ? 0 : cheapest;
    }

    public IReadOnlyList<PortDefinition> GetDestinationsFromPort(PortDefinition origin, IEnumerable<PortDefinition> ports)
    {
        return ports
            .Where(port => port.Id != origin.Id)
            .OrderBy(port => port.Zone)
            .ThenBy(port => port.Name, StringComparer.Ordinal)
            .ToList();
    }
}

File: internal/services/travel_service.go
package services

import (
	"sort"

	"github.com/oberones/star-smuggler-go/internal/domain"
)

type TravelService struct{}

type TravelQuote struct {
	Destination domain.PortDefinition
	Cost        int
}

func (s TravelService) GetTravelCost(fromPort domain.PortDefinition, toPort domain.PortDefinition) int {
	if fromPort.ID == toPort.ID {
		return 0
	}

	baseCost := 15
	zoneDifference := abs(zoneRank(fromPort.Zone) - zoneRank(toPort.Zone))
	cost := baseCost + (zoneDifference * 2)
	if zoneDifference >= 2 {
		cost *= 2
	}
	return cost
}

func (s TravelService) GetCheapestTravelCostFromPort(origin domain.PortDefinition, ports []domain.PortDefinition) int {
	cheapest := 0
	for _, destination := range ports {
		if destination.ID == origin.ID {
			continue
		}

		cost := s.GetTravelCost(origin, destination)
		if cheapest == 0 || cost < cheapest {
			cheapest = cost
		}
	}
	return cheapest
}

func (s TravelService) GetDestinationsFromPort(origin domain.PortDefinition, ports []domain.PortDefinition) []domain.PortDefinition {
	destinations := make([]domain.PortDefinition, 0, len(ports))
	for _, port := range ports {
		if port.ID != origin.ID {
			destinations = append(destinations, port)
		}
	}

	sort.Slice(destinations, func(i, j int) bool {
		left := destinations[i]
		right := destinations[j]
		if zoneRank(left.Zone) == zoneRank(right.Zone) {
			return left.Name < right.Name
		}
		return zoneRank(left.Zone) < zoneRank(right.Zone)
	})

	return destinations
}

func zoneRank(zone domain.PortZone) int {
	switch zone {
	case domain.PortZoneInner:
		return 0
	case domain.PortZoneOuter:
		return 1
	case domain.PortZoneFringe:
		return 2
	default:
		return 0
	}
}

func abs(value int) int {
	if value < 0 {
		return -value
	}
	return value
}
